File: leetcode/easy/94-inorder/recursive.cpp
/**
 * @file recursive.cpp
 * @author Vortexx2
 * @brief Problem 94 - Binary Tree Inorder Traversal
 * @date 21-07-2021
 * 
 * Runtime - 0 ms 
 * Memory Usage - 8.3 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  vector<int> inorderTraversal(TreeNode *root)
  {
    vector<int> res;
    helper(root, res);
    return res;
  }

  void helper(TreeNode *root, vector<int> &arr)
  {
    if (root)
    {
      helper(root->left, arr);
      arr.push_back(root->val);
      helper(root->right, arr);
    }
  }
};

File: leetcode/easy/94-inorder/recursive.py
"""
  @author Vortexx2
  Problem 94 - Binary Tree Inorder Traversal
  
  Runtime - 32 ms
  Memory Usage - 14.3 MB
"""
from typing import List


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def inorderTraversal(self, root: TreeNode) -> List[int]:
    res = []
    self.helper(root, res)
    return res

  def helper(self, root: TreeNode, arr: List[int]) -> None:
    if root:
      self.helper(root.left, arr)
      arr.append(root.val)
      self.helper(root.right, arr)
